File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 29/05/2025, 21:43:00
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')': '(', '}': '{', ']': '['}
        for char in s:
            if char in mapping.values():
                stack.append(char)
            else:
                if not stack or stack.pop() != mapping[char]:
                    return False
        return not stack

File: 20-ValidParentheses/20-ValidParentheses.js
// Last updated: 29/05/2025, 21:43:03
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = []
    for (let i = 0; i < s.length; i++) {
        if (stack.at(-1) === s[i]) {
            stack.pop()
        } else if (s[i] === '(') {
            stack.push(')')
        } else if (s[i] === '[') {
            stack.push(']')
        } else if (s[i] === '{') {
            stack.push('}')
        } else {
            return false
        }
    }
    return !stack.length
};